File: clients/go/pg_mentat_client.go
// Package pgmentat provides a direct PostgreSQL client for pg_mentat.
//
// No mentatd daemon required. Connect directly to PostgreSQL and call
// the pg_mentat extension functions via standard SQL.
//
// Requirements:
//
//	go get github.com/jackc/pgx/v5
//
// Usage:
//
//	client, _ := pgmentat.New(ctx, "postgresql://localhost/postgres")
//	defer client.Close()
//	client.Transact(ctx, `[{:db/ident :person/name ...}]`)
//	results, _ := client.Query(ctx, `[:find ?name :where [?e :person/name ?name]]`, nil)
package pgmentat

import (
	"context"
	"encoding/json"
	"fmt"

	"github.com/jackc/pgx/v5/pgxpool"
)

// MentatError is returned when a pg_mentat SQL function call fails.
type MentatError struct {
	Message string
}

func (e *MentatError) Error() string {
	return fmt.Sprintf("pg_mentat: %s", e.Message)
}

// Client is a direct PostgreSQL client for pg_mentat.
type Client struct {
	pool     *pgxpool.Pool
	ownsPool bool
}

// New creates a client with a new connection pool.
func New(ctx context.Context, connString string) (*Client, error) {
	pool, err := pgxpool.New(ctx, connString)
	if err != nil {
		return nil, fmt.Errorf("pg_mentat: connect: %w", err)
	}
	return &Client{pool: pool, ownsPool: true}, nil
}

// NewFromPool creates a client that reuses an existing connection pool.
func NewFromPool(pool *pgxpool.Pool) *Client {
	return &Client{pool: pool, ownsPool: false}
}

// Close releases the connection pool (only if we created it).
func (c *Client) Close() {
	if c.ownsPool {
		c.pool.Close()
	}
}

// Transact executes an EDN transaction and returns the transaction report.
func (c *Client) Transact(ctx context.Context, ednTx string) (map[string]interface{}, error) {
	var result string
	err := c.pool.QueryRow(ctx, "SELECT mentat_transact($1)", ednTx).Scan(&result)
	if err != nil {
		return nil, fmt.Errorf("pg_mentat: transact: %w", err)
	}
	var report map[string]interface{}
	if err := json.Unmarshal([]byte(result), &report); err != nil {
		return nil, fmt.Errorf("pg_mentat: parse transact result: %w", err)
	}
	return report, nil
}

// Query executes a Datalog query. Pass nil for inputs if none are needed.
func (c *Client) Query(ctx context.Context, datalog string, inputs map[string]interface{}) (interface{}, error) {
	if inputs == nil {
		inputs = map[string]interface{}{}
	}
	inputsJSON, err := json.Marshal(inputs)
	if err != nil {
		return nil, fmt.Errorf("pg_mentat: marshal inputs: %w", err)
	}

	var resultJSON json.RawMessage
	err = c.pool.QueryRow(ctx,
		"SELECT mentat_query($1, $2::jsonb)",
		datalog, string(inputsJSON),
	).Scan(&resultJSON)
	if err != nil {
		return nil, fmt.Errorf("pg_mentat: query: %w", err)
	}

	var result interface{}
	if err := json.Unmarshal(resultJSON, &result); err != nil {
		return nil, fmt.Errorf("pg_mentat: parse query result: %w", err)
	}
	return result, nil
}

// Pull retrieves attributes for an entity using a pull pattern.
func (c *Client) Pull(ctx context.Context, pattern string, entityID int64) (map[string]interface{}, error) {
	var resultJSON json.RawMessage
	err := c.pool.QueryRow(ctx,
		"SELECT mentat_pull($1, $2)",
		pattern, entityID,
	).Scan(&resultJSON)
	if err != nil {
		return nil, fmt.Errorf("pg_mentat: pull: %w", err)
	}

	var result map[string]interface{}
	if err := json.Unmarshal(resultJSON, &result); err != nil {
		return nil, fmt.Errorf("pg_mentat: parse pull result: %w", err)
	}
	return result, nil
}

// PullMany retrieves attributes for multiple entities.
func (c *Client) PullMany(ctx context.Context, pattern string, entityIDs []int64) ([]map[string]interface{}, error) {
	var resultJSON json.RawMessage
	err := c.pool.QueryRow(ctx,
		"SELECT mentat_pull_many($1, $2)",
		pattern, entityIDs,
	).Scan(&resultJSON)
	if err != nil {
		return nil, fmt.Errorf("pg_mentat: pull_many: %w", err)
	}

	var result []map[string]interface{}
	if err := json.Unmarshal(resultJSON, &result); err != nil {
		return nil, fmt.Errorf("pg_mentat: parse pull_many result: %w", err)
	}
	return result, nil
}

// Entity returns all attributes of an entity.
func (c *Client) Entity(ctx context.Context, entityID int64) (map[string]interface{}, error) {
	var resultJSON json.RawMessage
	err := c.pool.QueryRow(ctx, "SELECT mentat_entity($1)", entityID).Scan(&resultJSON)
	if err != nil {
		return nil, fmt.Errorf("pg_mentat: entity: %w", err)
	}

	var result map[string]interface{}
	if err := json.Unmarshal(resultJSON, &result); err != nil {
		return nil, fmt.Errorf("pg_mentat: parse entity result: %w", err)
	}
	return result, nil
}

// Schema returns the current schema.
func (c *Client) Schema(ctx context.Context) (map[string]interface{}, error) {
	var resultJSON json.RawMessage
	err := c.pool.QueryRow(ctx, "SELECT mentat_schema()").Scan(&resultJSON)
	if err != nil {
		return nil, fmt.Errorf("pg_mentat: schema: %w", err)
	}

	var result map[string]interface{}
	if err := json.Unmarshal(resultJSON, &result); err != nil {
		return nil, fmt.Errorf("pg_mentat: parse schema result: %w", err)
	}
	return result, nil
}

// Explain returns the execution plan for a Datalog query.
func (c *Client) Explain(ctx context.Context, datalog string, inputs map[string]interface{}) (interface{}, error) {
	if inputs == nil {
		inputs = map[string]interface{}{}
	}
	inputsJSON, err := json.Marshal(inputs)
	if err != nil {
		return nil, fmt.Errorf("pg_mentat: marshal inputs: %w", err)
	}

	var resultJSON json.RawMessage
	err = c.pool.QueryRow(ctx,
		"SELECT mentat_explain($1, $2::jsonb)",
		datalog, string(inputsJSON),
	).Scan(&resultJSON)
	if err != nil {
		return nil, fmt.Errorf("pg_mentat: explain: %w", err)
	}

	var result interface{}
	if err := json.Unmarshal(resultJSON, &result); err != nil {
		return nil, fmt.Errorf("pg_mentat: parse explain result: %w", err)
	}
	return result, nil
}

// -- Native SQL view access ------------------------------------------

// schemaForStore derives the PostgreSQL schema name for a store.
func schemaForStore(store string) string {
	if store == "default" || store == "" {
		return "mentat"
	}
	return "mentat_" + store
}

// Fact represents a row from the facts view.
type Fact struct {
	EntityID  int64   `json:"entity_id"`
	Attribute string  `json:"attribute"`
	Value     string  `json:"value"`
	ValueType string  `json:"value_type"`
	Tx        int64   `json:"tx"`
	TxTime    *string `json:"tx_time,omitempty"`
}

// Facts queries the facts view with optional filters.
func (c *Client) Facts(ctx context.Context, store string, entityID *int64, attribute *string) ([]Fact, error) {
	schema := schemaForStore(store)
	sql := fmt.Sprintf("SELECT entity_id, attribute, value, value_type, tx, tx_time::TEXT FROM %s.facts", schema)
	args := []interface{}{}
	wheres := []string{}
	if entityID != nil {
		args = append(args, *entityID)
		wheres = append(wheres, fmt.Sprintf("entity_id = $%d", len(args)))
	}
	if attribute != nil {
		args = append(args, *attribute)
		wheres = append(wheres, fmt.Sprintf("attribute = $%d", len(args)))
	}
	if len(wheres) > 0 {
		sql += " WHERE " + wheres[0]
		for _, w := range wheres[1:] {
			sql += " AND " + w
		}
	}
	sql += " ORDER BY entity_id, attribute"

	rows, err := c.pool.Query(ctx, sql, args...)
	if err != nil {
		return nil, fmt.Errorf("pg_mentat: facts: %w", err)
	}
	defer rows.Close()

	var facts []Fact
	for rows.Next() {
		var f Fact
		if err := rows.Scan(&f.EntityID, &f.Attribute, &f.Value, &f.ValueType, &f.Tx, &f.TxTime); err != nil {
			return nil, fmt.Errorf("pg_mentat: scan fact: %w", err)
		}
		facts = append(facts, f)
	}
	return facts, rows.Err()
}

// Reference represents a row from the entity_references view.
type Reference struct {
	SourceEntity int64   `json:"source_entity"`
	Attribute    string  `json:"attribute"`
	TargetEntity int64   `json:"target_entity"`
	TargetIdent  *string `json:"target_ident,omitempty"`
	Tx           int64   `json:"tx"`
}

// EntityReferences queries the entity_references view with optional filters.
func (c *Client) EntityReferences(ctx context.Context, store string, source *int64, target *int64) ([]Reference, error) {
	schema := schemaForStore(store)
	sql := fmt.Sprintf("SELECT source_entity, attribute, target_entity, target_ident, tx FROM %s.entity_references", schema)
	args := []interface{}{}
	wheres := []string{}
	if source != nil {
		args = append(args, *source)
		wheres = append(wheres, fmt.Sprintf("source_entity = $%d", len(args)))
	}
	if target != nil {
		args = append(args, *target)
		wheres = append(wheres, fmt.Sprintf("target_entity = $%d", len(args)))
	}
	if len(wheres) > 0 {
		sql += " WHERE " + wheres[0]
		for _, w := range wheres[1:] {
			sql += " AND " + w
		}
	}

	rows, err := c.pool.Query(ctx, sql, args...)
	if err != nil {
		return nil, fmt.Errorf("pg_mentat: entity_references: %w", err)
	}
	defer rows.Close()

	var refs []Reference
	for rows.Next() {
		var r Reference
		if err := rows.Scan(&r.SourceEntity, &r.Attribute, &r.TargetEntity, &r.TargetIdent, &r.Tx); err != nil {
			return nil, fmt.Errorf("pg_mentat: scan reference: %w", err)
		}
		refs = append(refs, r)
	}
	return refs, rows.Err()
}

// HistoryEntry represents a row from the entity_history view.
type HistoryEntry struct {
	EntityID  int64   `json:"entity_id"`
	Attribute string  `json:"attribute"`
	Value     string  `json:"value"`
	ValueType string  `json:"value_type"`
	Tx        int64   `json:"tx"`
	TxTime    *string `json:"tx_time,omitempty"`
	Operation string  `json:"operation"`
}

// EntityHistory queries the entity_history view.
func (c *Client) EntityHistory(ctx context.Context, store string, entityID *int64) ([]HistoryEntry, error) {
	schema := schemaForStore(store)
	sql := fmt.Sprintf("SELECT entity_id, attribute, value, value_type, tx, tx_time::TEXT, operation FROM %s.entity_history", schema)
	args := []interface{}{}
	if entityID != nil {
		args = append(args, *entityID)
		sql += fmt.Sprintf(" WHERE entity_id = $%d", len(args))
	}
	sql += " ORDER BY tx DESC"

	rows, err := c.pool.Query(ctx, sql, args...)
	if err != nil {
		return nil, fmt.Errorf("pg_mentat: entity_history: %w", err)
	}
	defer rows.Close()

	var entries []HistoryEntry
	for rows.Next() {
		var e HistoryEntry
		if err := rows.Scan(&e.EntityID, &e.Attribute, &e.Value, &e.ValueType, &e.Tx, &e.TxTime, &e.Operation); err != nil {
			return nil, fmt.Errorf("pg_mentat: scan history: %w", err)
		}
		entries = append(entries, e)
	}
	return entries, rows.Err()
}

// TxLogEntry represents a row from the tx_log view.
type TxLogEntry struct {
	Tx         int64   `json:"tx"`
	TxTime     *string `json:"tx_time,omitempty"`
	DatomCount int64   `json:"datom_count"`
}

// TxLog queries the tx_log view.
func (c *Client) TxLog(ctx context.Context, store string, limit int) ([]TxLogEntry, error) {
	if limit <= 0 {
		limit = 100
	}
	schema := schemaForStore(store)
	sql := fmt.Sprintf("SELECT tx, tx_time::TEXT, datom_count FROM %s.tx_log ORDER BY tx DESC LIMIT $1", schema)

	rows, err := c.pool.Query(ctx, sql, limit)
	if err != nil {
		return nil, fmt.Errorf("pg_mentat: tx_log: %w", err)
	}
	defer rows.Close()

	var entries []TxLogEntry
	for rows.Next() {
		var e TxLogEntry
		if err := rows.Scan(&e.Tx, &e.TxTime, &e.DatomCount); err != nil {
			return nil, fmt.Errorf("pg_mentat: scan tx_log: %w", err)
		}
		entries = append(entries, e)
	}
	return entries, rows.Err()
}

// LookupEntity finds entities by attribute value.
func (c *Client) LookupEntity(ctx context.Context, store string, attribute string, value string) ([]int64, error) {
	schema := schemaForStore(store)
	sql := fmt.Sprintf("SELECT entity_id FROM %s.lookup_entity($1, $2)", schema)

	rows, err := c.pool.Query(ctx, sql, attribute, value)
	if err != nil {
		return nil, fmt.Errorf("pg_mentat: lookup_entity: %w", err)
	}
	defer rows.Close()

	var ids []int64
	for rows.Next() {
		var id int64
		if err := rows.Scan(&id); err != nil {
			return nil, fmt.Errorf("pg_mentat: scan lookup: %w", err)
		}
		ids = append(ids, id)
	}
	return ids, rows.Err()
}

// EntityValue gets a single attribute value for an entity.
func (c *Client) EntityValue(ctx context.Context, store string, entityID int64, attribute string) (*string, error) {
	schema := schemaForStore(store)
	sql := fmt.Sprintf("SELECT %s.entity_value($1, $2)", schema)

	var result *string
	err := c.pool.QueryRow(ctx, sql, entityID, attribute).Scan(&result)
	if err != nil {
		return nil, fmt.Errorf("pg_mentat: entity_value: %w", err)
	}
	return result, nil
}

// TextSearchResult represents a full-text search match.
type TextSearchResult struct {
	EntityID  int64   `json:"entity_id"`
	Attribute string  `json:"attribute"`
	Value     string  `json:"value"`
	Rank      float32 `json:"rank"`
}

// FindText performs full-text search across all text values.
func (c *Client) FindText(ctx context.Context, store string, searchQuery string) ([]TextSearchResult, error) {
	schema := schemaForStore(store)
	sql := fmt.Sprintf("SELECT entity_id, attribute, value, rank FROM %s.find_text($1)", schema)

	rows, err := c.pool.Query(ctx, sql, searchQuery)
	if err != nil {
		return nil, fmt.Errorf("pg_mentat: find_text: %w", err)
	}
	defer rows.Close()

	var results []TextSearchResult
	for rows.Next() {
		var r TextSearchResult
		if err := rows.Scan(&r.EntityID, &r.Attribute, &r.Value, &r.Rank); err != nil {
			return nil, fmt.Errorf("pg_mentat: scan text search: %w", err)
		}
		results = append(results, r)
	}
	return results, rows.Err()
}

File: clients/rust/pg_mentat_client.rs
//! pg_mentat Rust client -- Direct PostgreSQL access.
//!
//! No mentatd daemon required. Connect directly to PostgreSQL and call
//! the pg_mentat extension functions via standard SQL.
//!
//! # Dependencies (Cargo.toml)
//!
//! ```toml
//! [dependencies]
//! tokio-postgres = "0.7"
//! serde_json = "1"
//! tokio = { version = "1", features = ["full"] }
//! deadpool-postgres = "0.14"  # optional, for connection pooling
//! ```
//!
//! # Usage
//!
//! ```rust,no_run
//! use pg_mentat_client::MentatClient;
//!
//! #[tokio::main]
//! async fn main() {
//!     let client = MentatClient::connect("host=localhost dbname=postgres").await.unwrap();
//!     client.transact(r#"[{:db/ident :person/name ...}]"#).await.unwrap();
//!     let results = client.query("[:find ?name :where [?e :person/name ?name]]", None).await.unwrap();
//! }
//! ```

use serde_json::Value;
use std::error::Error;
use std::fmt;
use tokio_postgres::{Client, NoTls};

/// Error from a pg_mentat SQL function call.
#[derive(Debug)]
pub struct MentatError {
    pub message: String,
}

impl fmt::Display for MentatError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "pg_mentat: {}", self.message)
    }
}

impl Error for MentatError {}

/// Direct PostgreSQL client for pg_mentat.
///
/// Calls mentat_transact(), mentat_query(), mentat_pull(), mentat_entity(),
/// and mentat_schema() as SQL functions -- no HTTP daemon needed.
pub struct MentatClient {
    client: Client,
    // The connection task handle; must be kept alive.
    _handle: tokio::task::JoinHandle<()>,
}

impl MentatClient {
    /// Connect to PostgreSQL.
    ///
    /// `config` is a libpq-style connection string, e.g.
    /// `"host=localhost dbname=postgres"`.
    pub async fn connect(config: &str) -> Result<Self, Box<dyn Error>> {
        let (client, connection) = tokio_postgres::connect(config, NoTls).await?;
        let handle = tokio::spawn(async move {
            if let Err(e) = connection.await {
                eprintln!("pg_mentat: connection error: {}", e);
            }
        });
        Ok(Self {
            client,
            _handle: handle,
        })
    }

    /// Execute an EDN transaction. Returns the transaction report as JSON.
    pub async fn transact(&self, edn_tx: &str) -> Result<Value, Box<dyn Error>> {
        let row = self
            .client
            .query_one("SELECT mentat_transact($1)", &[&edn_tx])
            .await?;
        let result: String = row.get(0);
        Ok(serde_json::from_str(&result)?)
    }

    /// Execute a Datalog query.
    pub async fn query(
        &self,
        datalog: &str,
        inputs: Option<&Value>,
    ) -> Result<Value, Box<dyn Error>> {
        let empty = serde_json::json!({});
        let inputs_json = serde_json::to_string(inputs.unwrap_or(&empty))?;
        let row = self
            .client
            .query_one(
                "SELECT mentat_query($1, $2::jsonb)",
                &[&datalog, &inputs_json],
            )
            .await?;
        let result: Value = row.get(0);
        Ok(result)
    }

    /// Pull attributes for an entity.
    pub async fn pull(
        &self,
        pattern: &str,
        entity_id: i64,
    ) -> Result<Value, Box<dyn Error>> {
        let row = self
            .client
            .query_one("SELECT mentat_pull($1, $2)", &[&pattern, &entity_id])
            .await?;
        let result: Value = row.get(0);
        Ok(result)
    }

    /// Pull attributes for multiple entities.
    pub async fn pull_many(
        &self,
        pattern: &str,
        entity_ids: &[i64],
    ) -> Result<Value, Box<dyn Error>> {
        let row = self
            .client
            .query_one(
                "SELECT mentat_pull_many($1, $2)",
                &[&pattern, &entity_ids],
            )
            .await?;
        let result: Value = row.get(0);
        Ok(result)
    }

    /// Get all attributes of an entity.
    pub async fn entity(&self, entity_id: i64) -> Result<Value, Box<dyn Error>> {
        let row = self
            .client
            .query_one("SELECT mentat_entity($1)", &[&entity_id])
            .await?;
        let result: Value = row.get(0);
        Ok(result)
    }

    /// Return the current schema.
    pub async fn schema(&self) -> Result<Value, Box<dyn Error>> {
        let row = self.client.query_one("SELECT mentat_schema()", &[]).await?;
        let result: Value = row.get(0);
        Ok(result)
    }

    /// Return the query execution plan.
    pub async fn explain(
        &self,
        datalog: &str,
        inputs: Option<&Value>,
    ) -> Result<Value, Box<dyn Error>> {
        let empty = serde_json::json!({});
        let inputs_json = serde_json::to_string(inputs.unwrap_or(&empty))?;
        let row = self
            .client
            .query_one(
                "SELECT mentat_explain($1, $2::jsonb)",
                &[&datalog, &inputs_json],
            )
            .await?;
        let result: Value = row.get(0);
        Ok(result)
    }

    // -- Native SQL view access ------------------------------------------

    /// Query the facts view for human-readable EAVT data.
    ///
    /// Returns rows as JSON arrays. Optionally filter by entity_id and/or
    /// attribute ident.
    pub async fn facts(
        &self,
        entity_id: Option<i64>,
        attribute: Option<&str>,
        store: &str,
    ) -> Result<Vec<Value>, Box<dyn Error>> {
        let schema = Self::schema_for_store(store);
        let mut sql = format!(
            "SELECT row_to_json(f) FROM (SELECT entity_id, attribute, value, value_type, tx, tx_time FROM {}.facts",
            schema
        );
        let mut params: Vec<Box<dyn tokio_postgres::types::ToSql + Sync>> = Vec::new();
        let mut wheres = Vec::new();

        if let Some(eid) = entity_id {
            params.push(Box::new(eid));
            wheres.push(format!("entity_id = ${}", params.len()));
        }
        if let Some(attr) = attribute {
            params.push(Box::new(attr.to_string()));
            wheres.push(format!("attribute = ${}", params.len()));
        }
        if !wheres.is_empty() {
            sql.push_str(" WHERE ");
            sql.push_str(&wheres.join(" AND "));
        }
        sql.push_str(" ORDER BY entity_id, attribute) f");

        let param_refs: Vec<&(dyn tokio_postgres::types::ToSql + Sync)> =
            params.iter().map(|p| p.as_ref()).collect();
        let rows = self.client.query(&sql, &param_refs).await?;
        Ok(rows.iter().map(|r| r.get::<_, Value>(0)).collect())
    }

    /// Query entity_references view for relationship navigation.
    pub async fn entity_references(
        &self,
        source: Option<i64>,
        target: Option<i64>,
        store: &str,
    ) -> Result<Vec<Value>, Box<dyn Error>> {
        let schema = Self::schema_for_store(store);
        let mut sql = format!(
            "SELECT row_to_json(r) FROM (SELECT source_entity, attribute, target_entity, target_ident, tx FROM {}.entity_references",
            schema
        );
        let mut params: Vec<Box<dyn tokio_postgres::types::ToSql + Sync>> = Vec::new();
        let mut wheres = Vec::new();

        if let Some(s) = source {
            params.push(Box::new(s));
            wheres.push(format!("source_entity = ${}", params.len()));
        }
        if let Some(t) = target {
            params.push(Box::new(t));
            wheres.push(format!("target_entity = ${}", params.len()));
        }
        if !wheres.is_empty() {
            sql.push_str(" WHERE ");
            sql.push_str(&wheres.join(" AND "));
        }
        sql.push_str(") r");

        let param_refs: Vec<&(dyn tokio_postgres::types::ToSql + Sync)> =
            params.iter().map(|p| p.as_ref()).collect();
        let rows = self.client.query(&sql, &param_refs).await?;
        Ok(rows.iter().map(|r| r.get::<_, Value>(0)).collect())
    }

    /// Query entity_history view for temporal data.
    pub async fn entity_history(
        &self,
        entity_id: Option<i64>,
        store: &str,
    ) -> Result<Vec<Value>, Box<dyn Error>> {
        let schema = Self::schema_for_store(store);
        let mut sql = format!(
            "SELECT row_to_json(h) FROM (SELECT entity_id, attribute, value, value_type, tx, tx_time, operation FROM {}.entity_history",
            schema
        );
        let mut params: Vec<Box<dyn tokio_postgres::types::ToSql + Sync>> = Vec::new();

        if let Some(eid) = entity_id {
            params.push(Box::new(eid));
            sql.push_str(&format!(" WHERE entity_id = ${}", params.len()));
        }
        sql.push_str(" ORDER BY tx DESC) h");

        let param_refs: Vec<&(dyn tokio_postgres::types::ToSql + Sync)> =
            params.iter().map(|p| p.as_ref()).collect();
        let rows = self.client.query(&sql, &param_refs).await?;
        Ok(rows.iter().map(|r| r.get::<_, Value>(0)).collect())
    }

    /// Query tx_log view for transaction history.
    pub async fn tx_log(
        &self,
        limit: i64,
        store: &str,
    ) -> Result<Vec<Value>, Box<dyn Error>> {
        let schema = Self::schema_for_store(store);
        let sql = format!(
            "SELECT row_to_json(t) FROM (SELECT tx, tx_time, datom_count FROM {}.tx_log ORDER BY tx DESC LIMIT $1) t",
            schema
        );
        let rows = self.client.query(&sql, &[&limit]).await?;
        Ok(rows.iter().map(|r| r.get::<_, Value>(0)).collect())
    }

    /// Find entities by attribute value using the lookup_entity function.
    pub async fn lookup_entity(
        &self,
        attribute: &str,
        value: &str,
        store: &str,
    ) -> Result<Vec<i64>, Box<dyn Error>> {
        let schema = Self::schema_for_store(store);
        let sql = format!(
            "SELECT entity_id FROM {}.lookup_entity($1, $2)",
            schema
        );
        let rows = self
            .client
            .query(&sql, &[&attribute, &value])
            .await?;
        Ok(rows.iter().map(|r| r.get::<_, i64>(0)).collect())
    }

    /// Get a single attribute value for an entity, returned as text.
    pub async fn entity_value(
        &self,
        entity_id: i64,
        attribute: &str,
        store: &str,
    ) -> Result<Option<String>, Box<dyn Error>> {
        let schema = Self::schema_for_store(store);
        let sql = format!("SELECT {}.entity_value($1, $2)", schema);
        let row = self
            .client
            .query_one(&sql, &[&entity_id, &attribute])
            .await?;
        Ok(row.get::<_, Option<String>>(0))
    }

    /// Full-text search across all text values.
    pub async fn find_text(
        &self,
        search_query: &str,
        store: &str,
    ) -> Result<Vec<Value>, Box<dyn Error>> {
        let schema = Self::schema_for_store(store);
        let sql = format!(
            "SELECT row_to_json(r) FROM (SELECT entity_id, attribute, value, rank FROM {}.find_text($1)) r",
            schema
        );
        let rows = self.client.query(&sql, &[&search_query]).await?;
        Ok(rows.iter().map(|r| r.get::<_, Value>(0)).collect())
    }

    /// Derive the PostgreSQL schema name for a store.
    fn schema_for_store(store: &str) -> String {
        if store == "default" {
            "mentat".to_string()
        } else {
            format!("mentat_{}", store)
        }
    }
}

// ---------------------------------------------------------------------------
// Example usage
// ---------------------------------------------------------------------------
#[cfg(feature = "example")]
#[tokio::main]
async fn main() -> Result<(), Box<dyn Error>> {
    let dsn = std::env::var("PG_MENTAT_DSN")
        .unwrap_or_else(|_| "host=localhost dbname=postgres".into());

    let client = MentatClient::connect(&dsn).await?;

    // Define schema
    client
        .transact(
            r#"[
      {:db/ident :person/name
       :db/valueType :db.type/string
       :db/cardinality :db.cardinality/one}
      {:db/ident :person/email
       :db/valueType :db.type/string
       :db/cardinality :db.cardinality/one
       :db/unique :db.unique/identity}
    ]"#,
        )
        .await?;

    // Transact data
    client
        .transact(
            r#"[
      {:person/name "Alice" :person/email "alice@example.com"}
      {:person/name "Bob"   :person/email "bob@example.com"}
    ]"#,
        )
        .await?;

    // Query
    let results = client
        .query(
            r#"[:find ?name ?email
                :where
                [?e :person/name ?name]
                [?e :person/email ?email]]"#,
            None,
        )
        .await?;
    println!("Query results: {}", serde_json::to_string_pretty(&results)?);

    // Schema
    let schema = client.schema().await?;
    println!("Schema: {}", serde_json::to_string_pretty(&schema)?);

    Ok(())
}
